File: src/main/java/org/moinex/ui/dialog/wallettransaction/EditTransferController.java
/*
 * Filename: AddTransferController.java
 * Created on: October  4, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog.wallettransaction;

import jakarta.persistence.EntityNotFoundException;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.Objects;
import javafx.fxml.FXML;
import javafx.scene.control.Label;
import javafx.stage.Stage;
import lombok.NoArgsConstructor;
import org.moinex.error.MoinexException;
import org.moinex.model.Category;
import org.moinex.model.wallettransaction.Transfer;
import org.moinex.model.wallettransaction.Wallet;
import org.moinex.service.CalculatorService;
import org.moinex.service.CategoryService;
import org.moinex.service.WalletService;
import org.moinex.service.WalletTransactionService;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.moinex.util.WindowUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.stereotype.Controller;

/**
 * Controller for the Add Transfer dialog
 */
@Controller
@NoArgsConstructor
public class EditTransferController extends BaseTransferManagement {
    private Transfer oldTransfer;

    /**
     * Constructor
     *
     * @param walletService            WalletService
     * @param walletTransactionService WalletTransactionService
     * @param calculatorService        CalculatorService
     * @note This constructor is used for dependency injection
     */
    @Autowired
    public EditTransferController(
            WalletService walletService,
            WalletTransactionService walletTransactionService,
            CalculatorService calculatorService,
            CategoryService categoryService,
            ConfigurableApplicationContext springContext) {
        super(
                walletService,
                walletTransactionService,
                calculatorService,
                categoryService,
                springContext);
    }

    public void setTransfer(Transfer transfer) {
        if (transfer == null) {
            throw new IllegalArgumentException("Transfer cannot be null");
        }

        disableTransferValueListener();
        suggestionsHandler.disable();

        Wallet senderWallet = transfer.getSenderWallet();
        Wallet receiverWallet = transfer.getReceiverWallet();
        BigDecimal value = transfer.getAmount();
        String description = transfer.getDescription();
        LocalDate transferDate = LocalDate.from(transfer.getDate());
        Category category = transfer.getCategory();

        try {
            senderWalletComboBox.setValue(senderWallet);
            receiverWalletComboBox.setValue(receiverWallet);
            transferValueField.setText(value.toPlainString());
            descriptionField.setText(description);
            transferDatePicker.setValue(transferDate);
            categoryComboBox.setValue(category);

            senderWalletComboBox.setValue(senderWallet);
            updateSenderWalletBalance();

            receiverWalletComboBox.setValue(receiverWallet);
            updateReceiverWalletBalance();
        } finally {
            enableTransferValueListener();
            suggestionsHandler.enable();
        }

        oldTransfer = transfer;
    }

    @Override
    @FXML
    protected void initialize() {
        super.initialize();
    }

    @Override
    @FXML
    protected void handleSave() {
        Wallet senderWt = senderWalletComboBox.getValue();
        Wallet receiverWt = receiverWalletComboBox.getValue();
        String transferValueString = transferValueField.getText();
        String description = descriptionField.getText();
        LocalDate transferDate = transferDatePicker.getValue();

        Category category = categoryComboBox.getValue();

        if (senderWt == null
                || receiverWt == null
                || transferValueString == null
                || transferValueString.isBlank()
                || description == null
                || description.isBlank()
                || transferDate == null) {
            WindowUtils.showInformationDialog(
                    "Empty fields", "Please fill all required fields before saving");
            return;
        }

        try {
            BigDecimal transferValue = new BigDecimal(transferValueString);

            LocalTime currentTime = LocalTime.now();
            LocalDateTime dateTimeWithCurrentHour = transferDate.atTime(currentTime);

            // Check if it has any modification
            if (oldTransfer.getSenderWallet().equals(senderWt)
                    && oldTransfer.getReceiverWallet().equals(receiverWt)
                    && oldTransfer.getAmount().compareTo(transferValue) == 0
                    && oldTransfer.getDescription().equals(description)
                    && Objects.equals(
                            oldTransfer.getCategory() != null
                                    ? oldTransfer.getCategory().getId()
                                    : null,
                            category != null ? category.getId() : null)
                    && LocalDate.from(oldTransfer.getDate()).equals(transferDate)) {
                WindowUtils.showInformationDialog(
                        "No changes made", "No changes were made to the transfer.");
                return;
            } else // If there is any modification, update the transaction
            {
                oldTransfer.setAmount(transferValue);
                oldTransfer.setSenderWallet(senderWt);
                oldTransfer.setReceiverWallet(receiverWt);
                oldTransfer.setDescription(description);
                oldTransfer.setCategory(category);
                oldTransfer.setDate(dateTimeWithCurrentHour);
                walletTransactionService.updateTransfer(oldTransfer);

                WindowUtils.showSuccessDialog(
                        "Transfer created", "The transfer was successfully created.");
            }
        } catch (NumberFormatException e) {
            WindowUtils.showErrorDialog(
                    "Invalid transfer value", "Transfer value must be a number.");
        } catch (MoinexException.SameSourceDestinationException
                | IllegalArgumentException
                | EntityNotFoundException
                | IllegalStateException
                | MoinexException.InsufficientResourcesException
                | MoinexException.TransferFromMasterToVirtualWalletException e) {
            WindowUtils.showErrorDialog("Error while creating transfer", e.getMessage());
        }

        Stage stage = (Stage) descriptionField.getScene().getWindow();
        stage.close();
    }

    @Override
    protected void updateAfterBalance(
            Wallet currentWallet, Wallet otherWallet, Label label, boolean isSender) {

        String transferValueString = transferValueField.getText();

        if (transferValueString == null
                || transferValueString.isBlank()
                || currentWallet == null
                || currentWallet.equals(otherWallet)) {
            UIUtils.resetLabel(label);
            return;
        }

        try {
            BigDecimal newAmount =
                    new BigDecimal(transferValueString).setScale(2, RoundingMode.HALF_UP);
            if (newAmount.compareTo(BigDecimal.ZERO) < 0) {
                UIUtils.resetLabel(label);
                return;
            }

            BigDecimal balance = currentWallet.getBalance();
            BigDecimal oldAmount = oldTransfer.getAmount();
            Wallet oldSender = oldTransfer.getSenderWallet();
            Wallet oldReceiver = oldTransfer.getReceiverWallet();
            boolean sameWallet =
                    (isSender && currentWallet.equals(oldSender))
                            || (!isSender && currentWallet.equals(oldReceiver));

            BigDecimal afterBalance;

            if (sameWallet) {
                BigDecimal diff = newAmount.subtract(oldAmount);

                if (isSender) {
                    afterBalance = balance.subtract(diff);
                } else {
                    afterBalance = balance.add(diff);
                }
            } else {
                if (isSender) {
                    if (currentWallet.isMaster()
                            && otherWallet.isVirtual()
                            && otherWallet.getMasterWallet().equals(currentWallet)) {
                        WindowUtils.showInformationDialog(
                                "Invalid Transfer",
                                "Cannot transfer from a master wallet to its virtual wallet.");
                        return;
                    }
                    afterBalance = balance.subtract(newAmount);
                } else {
                    if (otherWallet != null
                            && otherWallet.isVirtual()
                            && otherWallet.getMasterWallet().equals(currentWallet)) {
                        afterBalance = balance;
                    } else {
                        afterBalance = balance.add(newAmount);
                    }
                }
            }

            if (afterBalance.compareTo(BigDecimal.ZERO) < 0) {
                UIUtils.setLabelStyle(label, Constants.NEGATIVE_BALANCE_STYLE);
            } else {
                UIUtils.setLabelStyle(label, Constants.NEUTRAL_BALANCE_STYLE);
            }

            label.setText(UIUtils.formatCurrency(afterBalance));

        } catch (NumberFormatException e) {
            UIUtils.resetLabel(label);
        }
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/wallettransaction/update/EditTransferController.kt
/*
 * Filename: EditTransferController.kt (original filename: AddTransferController.java)
 * Created on: October  4, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 21/03/2026
 */

package org.moinex.ui.dialog.wallettransaction.update

import jakarta.persistence.EntityNotFoundException
import javafx.fxml.FXML
import javafx.scene.control.Label
import javafx.stage.Stage
import org.moinex.common.constant.Styles
import org.moinex.common.constant.TranslationKeys
import org.moinex.common.extension.isEqual
import org.moinex.common.extension.toRounded
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.exception.MoinexException
import org.moinex.model.wallettransaction.Transfer
import org.moinex.model.wallettransaction.Wallet
import org.moinex.service.CalculatorService
import org.moinex.service.CategoryService
import org.moinex.service.PreferencesService
import org.moinex.service.wallet.WalletService
import org.moinex.ui.dialog.wallettransaction.base.BaseTransferManagement
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.stereotype.Controller
import java.math.BigDecimal
import java.time.LocalDate
import java.time.LocalTime

@Controller
class EditTransferController(
    walletService: WalletService,
    calculatorService: CalculatorService,
    categoryService: CategoryService,
    preferencesService: PreferencesService,
    springContext: ConfigurableApplicationContext,
) : BaseTransferManagement(
        walletService,
        calculatorService,
        categoryService,
        preferencesService,
        springContext,
    ) {
    private var oldTransfer: Transfer? = null

    fun setTransfer(transfer: Transfer) {
        disableTransferValueListener()
        suggestionsHandler.disable()

        val senderWallet = transfer.senderWallet
        val receiverWallet = transfer.receiverWallet
        val value = transfer.amount
        val description = transfer.description
        val transferDate = LocalDate.from(transfer.date)
        val category = transfer.category

        senderWalletComboBox.value = senderWallet
        receiverWalletComboBox.value = receiverWallet
        transferValueField.text = value.toPlainString()
        descriptionField.text = description
        transferDatePicker.value = transferDate
        categoryComboBox.value = category

        senderWalletComboBox.value = senderWallet
        updateSenderWalletBalance()

        receiverWalletComboBox.value = receiverWallet
        updateReceiverWalletBalance()
        enableTransferValueListener()
        suggestionsHandler.enable()

        oldTransfer = transfer
    }

    @FXML
    override fun initialize() {
        super.initialize()
    }

    @FXML
    override fun handleSave() {
        val senderWt = senderWalletComboBox.value
        val receiverWt = receiverWalletComboBox.value
        val transferValueString = transferValueField.text
        val description = descriptionField.text
        val transferDate = transferDatePicker.value
        val category = categoryComboBox.value

        if (senderWt == null ||
            receiverWt == null ||
            transferValueString.isNullOrBlank() ||
            description.isNullOrBlank() ||
            transferDate == null
        ) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_EMPTY_FIELDS_TITLE),
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_EMPTY_FIELDS_MESSAGE),
            )
            return
        }

        runCatching {
            val transferValue = transferValueString.toBigDecimal()
            val currentTime = LocalTime.now()
            val dateTimeWithCurrentHour = transferDate.atTime(currentTime)

            if (oldTransfer!!.senderWallet == senderWt &&
                oldTransfer!!.receiverWallet == receiverWt &&
                oldTransfer!!.amount.isEqual(transferValue) &&
                oldTransfer!!.description == description &&
                oldTransfer!!.category?.id!! == category?.id!! &&
                LocalDate.from(oldTransfer!!.date) == transferDate
            ) {
                WindowUtils.showInformationDialog(
                    preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_NO_CHANGES_MADE_TITLE),
                    preferencesService.translate(
                        TranslationKeys.WALLETTRANSACTION_DIALOG_NO_CHANGES_MADE_TRANSFER_MESSAGE,
                    ),
                )
                return
            }

            oldTransfer!!.amount = transferValue
            oldTransfer!!.senderWallet = senderWt
            oldTransfer!!.receiverWallet = receiverWt
            oldTransfer!!.description = description
            oldTransfer!!.category = category
            oldTransfer!!.date = dateTimeWithCurrentHour
            walletService.updateTransfer(oldTransfer!!)

            WindowUtils.showSuccessDialog(
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_TRANSFER_UPDATED_TITLE),
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_TRANSFER_UPDATED_MESSAGE),
            )
        }.onFailure { e ->
            when (e) {
                is NumberFormatException -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(
                            TranslationKeys.WALLETTRANSACTION_DIALOG_INVALID_TRANSFER_VALUE_TITLE,
                        ),
                        preferencesService.translate(
                            TranslationKeys.WALLETTRANSACTION_DIALOG_INVALID_TRANSFER_VALUE_MESSAGE,
                        ),
                    )
                }
                is MoinexException.SameSourceDestinationException,
                is IllegalArgumentException,
                is EntityNotFoundException,
                is IllegalStateException,
                is MoinexException.InsufficientResourcesException,
                is MoinexException.TransferFromMasterToVirtualWalletException,
                -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(
                            TranslationKeys.WALLETTRANSACTION_DIALOG_ERROR_UPDATING_TRANSFER_TITLE,
                        ),
                        e.message ?: "Unknown error",
                    )
                }
                else -> throw e
            }
        }

        (descriptionField.scene.window as Stage).close()
    }

    override fun updateAfterBalance(
        mainWallet: Wallet?,
        otherWallet: Wallet?,
        afterBalanceLabel: Label,
        isSender: Boolean,
    ) {
        val transferValueString = transferValueField.text

        if (transferValueString.isNullOrBlank() ||
            mainWallet == null ||
            mainWallet == otherWallet
        ) {
            UIUtils.resetLabel(afterBalanceLabel)
            return
        }

        runCatching {
            val newAmount = transferValueString.toBigDecimal().toRounded()
            if (newAmount < BigDecimal.ZERO) {
                UIUtils.resetLabel(afterBalanceLabel)
                return
            }

            val afterBalance = calculateAfterBalance(mainWallet, isSender, newAmount, otherWallet)

            val style =
                if (afterBalance < BigDecimal.ZERO) {
                    Styles.NEGATIVE_BALANCE_STYLE
                } else {
                    Styles.NEUTRAL_BALANCE_STYLE
                }

            UIUtils.setLabelStyle(afterBalanceLabel, style)
            afterBalanceLabel.text = UIUtils.formatCurrency(afterBalance)
        }.onFailure {
            UIUtils.resetLabel(afterBalanceLabel)
        }
    }

    private fun calculateAfterBalance(
        mainWallet: Wallet,
        isSender: Boolean,
        newAmount: BigDecimal,
        otherWallet: Wallet?,
    ): BigDecimal {
        val balance = mainWallet.balance
        val oldAmount = oldTransfer!!.amount
        val oldSender = oldTransfer!!.senderWallet
        val oldReceiver = oldTransfer!!.receiverWallet
        val sameWallet =
            (isSender && mainWallet == oldSender) ||
                (!isSender && mainWallet == oldReceiver)

        return if (sameWallet) {
            val diff = newAmount - oldAmount
            if (isSender) {
                balance - diff
            } else {
                balance + diff
            }
        } else {
            if (isSender) {
                if (mainWallet.isMaster() &&
                    otherWallet?.isVirtual() == true &&
                    otherWallet.masterWallet == mainWallet
                ) {
                    WindowUtils.showInformationDialog(
                        preferencesService.translate(
                            TranslationKeys.WALLETTRANSACTION_DIALOG_INVALID_TRANSFER_TITLE,
                        ),
                        preferencesService.translate(
                            TranslationKeys.WALLETTRANSACTION_DIALOG_INVALID_TRANSFER_MESSAGE,
                        ),
                    )
                    return BigDecimal.ZERO
                }
                balance - newAmount
            } else {
                if (otherWallet?.isVirtual() == true &&
                    otherWallet.masterWallet == mainWallet
                ) {
                    balance
                } else {
                    balance + newAmount
                }
            }
        }
    }
}
